File: hema/dl_operators_assignments/Operators and testing/operators/src/max_pooling.cpp
#include "max_pooling.h"
#include <cmath>
#include <limits>

// Constructor to initialize pooling parameters
MaxPool2D::MaxPool2D(int pool_size, int stride)
    : pool_size(pool_size), stride(stride) {}

// Forward pass for max pooling
std::vector<std::vector<std::vector<float>>> MaxPool2D::forward(
    const std::vector<std::vector<std::vector<float>>>& input) {

    int input_depth = input.size();
    int input_height = input[0].size();
    int input_width = input[0][0].size();

    // Calculate output dimensions
    int output_height = (input_height - pool_size) / stride + 1;
    int output_width = (input_width - pool_size) / stride + 1;

    // Initialize output with appropriate dimensions
    std::vector<std::vector<std::vector<float>>> output(
        input_depth,
        std::vector<std::vector<float>>(output_height, std::vector<float>(output_width, 0)));

    // Perform max pooling
    for (int d = 0; d < input_depth; ++d) {
        for (int h = 0; h < output_height; ++h) {
            for (int w = 0; w < output_width; ++w) {
                float max_val = -std::numeric_limits<float>::infinity();

                // Pooling window
                for (int i = 0; i < pool_size; ++i) {
                    for (int j = 0; j < pool_size; ++j) {
                        int in_h = h * stride + i;
                        int in_w = w * stride + j;
                        if (in_h < input_height && in_w < input_width) {
                            max_val = std::max(max_val, input[d][in_h][in_w]);
                        }
                    }
                }
                output[d][h][w] = max_val;
            }
        }
    }

    return output;
}

File: rohithk/deeplearning_operators_assignment/src/python_operators/maxpooling.py
import cv2
import numpy as np
import json


def max_pooling(image, pool_size=2, stride=2):
    """
    Performs max pooling on the input image.

    Parameters:
        image (numpy.ndarray): Input image array (grayscale or color).
        pool_size (int): Size of the pooling window (default is 2).
        stride (int): Stride of the pooling window (default is 2).

    Returns:
        numpy.ndarray: The max pooled image.
    """
    # Get the image dimensions.
    h, w = image.shape[:2]

    # Calculate dimensions for the pooled image.
    new_h = (h - pool_size) // stride + 1
    new_w = (w - pool_size) // stride + 1

    # Initialize the pooled image array depending on image type.
    if len(image.shape) == 3:
        pooled = np.zeros((new_h, new_w, image.shape[2]), dtype=image.dtype)
    else:
        pooled = np.zeros((new_h, new_w), dtype=image.dtype)

    # Slide the window and perform max pooling.
    for i in range(new_h):
        for j in range(new_w):
            window = image[
                i * stride : i * stride + pool_size, j * stride : j * stride + pool_size
            ]
            pooled[i, j] = np.max(window, axis=(0, 1))

    return pooled


# Load the image using OpenCV.
image_path = "input_image.png"  # Replace with your image file path.
image = cv2.imread(image_path)

if image is None:
    raise FileNotFoundError(f"Unable to load image at {image_path}")

# Optional: convert to grayscale if needed.
# image = cv2.cvtColor(image, cv2.COLOR_BGR2GRAY)

# Apply max pooling with a 2x2 window and stride of 2.
pooled_image = max_pooling(image, pool_size=2, stride=2)

# Convert the numpy array to a list for JSON serialization.
pooled_list = pooled_image.tolist()

# Write the pooled image data to a JSON file.
with open("pooled_output.json", "w") as json_file:
    json.dump(pooled_list, json_file)

# Optionally, display the original and pooled images.
